omod/src/main/java/org/openmrs/module/owa/web/controller/OwaRestController.java

```diff
@@ -143,4 +146,54 @@
             }
             return appList;
         }
-}
+	
+	@RequestMapping(value = "/rest/owa/installapp", method = RequestMethod.POST)
+            @ResponseBody
+            public List<App> install(@RequestParam("downloadUrl") String url, HttpServletRequest request, HttpServletResponse response) throws IOException {
+               List<App> appList = new ArrayList<>();
+               URL downloadUrl = new URL(url);
+               if(Context.hasPrivilege("Manage OWA")) {
+                    String message;
+                    HttpSession session = request.getSession();
+                    if (!url.isEmpty()) {
+                        InputStream inputStream = ModuleUtil.getURLStream(downloadUrl);
+                        log.warn("url pathname: " + downloadUrl.getPath());                                           
+                        String fileName = downloadUrl.getQuery().substring(downloadUrl.getQuery().lastIndexOf("=") + 1);
+                        File file = ModuleUtil.insertModuleFile(inputStream, fileName);
+                        try (ZipFile zip = new ZipFile(file)) {
+                            if (zip.size() == 0) {
+                                message = messageSourceService.getMessage("owa.blank_zip");
+                                log.warn("Zip file is empty");
+                                session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
+                                response.sendError(500, message);
+                            } else {
+                                ZipEntry entry = zip.getEntry("manifest.webapp");
+                                if (entry == null) {
+                                    message = messageSourceService.getMessage("owa.manifest_not_found");
+                                    log.warn("Manifest file could not be found in app");
+                                    file.delete();
+                                    session.setAttribute(WebConstants.OPENMRS_ERROR_ATTR, message);
+                                    response.sendError(500, message);
+                                } else {                          
+                                    String contextPath = request.getScheme() + "://" + request.getServerName() + ":"
+                                            + request.getServerPort() + request.getContextPath();
+                                    appManager.installApp(file, fileName, contextPath);
+                                    message = messageSourceService.getMessage("owa.app_installed");
+                                    session.setAttribute(WebConstants.OPENMRS_MSG_ATTR, message);
```

Comment:
shouldn't there be `file.delete()` in this else block?

Reply:
@mkiterian On line 180, the `appManager.installApp` has an underlying implementation that deletes the file to ensure there is no memory leakage.